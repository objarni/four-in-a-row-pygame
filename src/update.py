from src.constants import DROP_DELAY_MS, RED, YELLOW, ROWS, EMPTY, COLUMNS, BOARD_LEFT, BOARD_RIGHT, DISC_DIAMETER
from src.messages import LeftMouseDownAt, Tick, ColumnWasClicked, MouseMovedTo, LeftMouseUpAt
from src.states import StartScreenState, GameState, GameOverState


def update(model, msg, audio_api):
    begin = lambda *args: args[-1]
    def assign(o, a, v):
        o.__setattr__(a, v)

    case_of = {
        (StartScreenState, LeftMouseDownAt): lambda: begin(
                audio_api.stop_music(),
                GameState()),
        (StartScreenState, Tick): lambda: begin(
                (assign(model, 'time', msg.time)),
                (audio_api.play_music('music') if not model.music_playing else None),
                (assign(model, 'music_playing', True)),
                model)
    }

    # if isinstance(model, StartScreenState):
    #     if isinstance(msg, LeftMouseDownAt):
    #         audio_api.stop_music()
    #         return GameState()
    #     if isinstance(msg, Tick):
    #         if not model.music_playing:
    #             audio_api.play_music('music')
    #             model.music_playing = True
    #         model.time = msg.time
    #         return model


    for condition in case_of:
        stateClass, msgClass = condition
        if isinstance(model, stateClass) and isinstance(msg, msgClass):
            return case_of[condition]()

    if isinstance(model, GameState):
        return update_gamestate(model, msg, audio_api)
    if isinstance(model, GameOverState):
        if isinstance(msg, LeftMouseDownAt):
            return StartScreenState()

    return model


def update_gamestate(gamestate, msg, audio_api):
    if isinstance(msg, Tick):
        gamestate.time = msg.time
        if (
            gamestate.mouse_down_time
            and gamestate.time > gamestate.mouse_down_time + DROP_DELAY_MS
        ):
            gamestate.mouse_down_time = None
            gamestate = update(gamestate, ColumnWasClicked(convert_to_column(gamestate.mouse_pos[0])), audio_api)
    if isinstance(msg, MouseMovedTo):
        gamestate.mouse_pos = msg.pos
    if (
        isinstance(msg, LeftMouseDownAt)
        and convert_to_column(msg.pos[0]) is not None
    ):
        gamestate.mouse_down = msg.pos
        gamestate.mouse_down_time = gamestate.time
    if isinstance(msg, LeftMouseUpAt):
        gamestate.mouse_down_time = None
    if isinstance(msg, ColumnWasClicked):
        if gamestate.board[(msg.column, 0)] != EMPTY:
            audio_api.play_sound('blocked')
        else:
            audio_api.play_sound('drop')
            gamestate.juice = 10
            gamestate.board = place_brick(gamestate.board, gamestate.whos_turn_is_it, msg.column)
            gamestate.whos_turn_is_it = (gamestate.whos_turn_is_it + 1) % 2
            for color in [RED, YELLOW]:
                won = check_winning_state(gamestate.board, color)
                if won:
                    gamestate = GameOverState(winner=color, board=gamestate.board)
    return gamestate


def place_brick(board, color, column):
    log(f"Placing brick color {print_color(color)} in column {column}")
    for i in range(ROWS):
        y = ROWS - i - 1
        if board[(column, y)] == EMPTY:
            board[(column, y)] = color
            break
    return board


def print_color(color):
    return 'red' if color == RED else 'yellow'


def check_winning_state(board, color):
    for (x, y) in positions_in_print_order():
        for dir in [(0, 1), (1, 0), (1, 1), (-1, 1)]:
            cells = extract(board, (x, y), dir)
            if all(cell == color for cell in cells):
                log(f"Found 4-in-a-row at {x, y} direction {dir}")
                return True
    return False


def extract(board, pos, dir):
    cells = []
    for i in range(4):
        p = (pos[0] + dir[0] * i, pos[1] + dir[1] * i)
        cells.append(board[p])
    return cells


def positions_in_print_order():
    for y in range(ROWS):
        for x in range(COLUMNS):
            yield (x, y)


def convert_to_column(x):
    if x < BOARD_LEFT:
        return None
    if x > BOARD_RIGHT:
        return None
    return (x - BOARD_LEFT) // DISC_DIAMETER


def log(msg):
    pass
    # print(msg)