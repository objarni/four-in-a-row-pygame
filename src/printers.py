from src.constants import EMPTY, RED, YELLOW, ROWS, COLUMNS
from src.states import StartScreenState, GameOverState, GameState
from src.update import print_color


def print_model(model):
    state_string = model.__class__.__name__ + '\n'
    if isinstance(model, StartScreenState):
        state_string += f'{model.time=}\n'
        state_string += f'{model.music_playing=}\n'
    if isinstance(model, GameOverState):
        state_string += print_board(model.board)
        state_string += f'\n{print_color(model.winner).title()} won.\n'
    if isinstance(model, GameState):
        state_string += f'It is {print_color(model.whos_turn_is_it)}s turn.\n'
        state_string += f'{model.time=}\n'
        state_string += f'The mouse is at {model.mouse_pos}.\n'
        state_string += f'{model.mouse_down_time=}\n'
        state_string += print_board(model.board)
    return state_string


def print_board(board):
    board_string = ''
    symbols = {
        EMPTY: 'O',
        RED: 'R',
        YELLOW: 'Y'
    }
    board_string += "0 1 2 3 4 5 6\n"
    board_string += "-------------\n"
    for y in range(ROWS):
        board_string += ' '.join(symbols[board[(x, y)]] for x in range(COLUMNS)) + '\n'
    return board_string