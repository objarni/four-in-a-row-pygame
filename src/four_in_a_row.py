import pygame

from src.messages import LeftMouseDownAt, LeftMouseUpAt, MouseMovedTo, Tick
from src.printers import print_model
from src.states import StartScreenState
from src.constants import WIDTH, HEIGHT, FPS

from src.update import update
from src.view import view


def main():
    global api
    pygame.init()
    pygame.mixer.init()
    screen = pygame.display.set_mode((WIDTH, HEIGHT))
    pygame.display.set_caption("Four in a row")
    resource_folder = 'res'
    drawing_api = DrawingAPI(screen, resource_folder)
    audio_api = AudioAPI(resource_folder)
    mainloop(drawing_api, audio_api)
    pygame.quit()


def mainloop(drawing_api, audio_api):
    model = StartScreenState()
    view(model, drawing_api)
    pygame.display.update()
    clock = pygame.time.Clock()
    while True:
        old_model_repr = print_model(model)

        # A tick happens every time around the loop!
        model = update(model, Tick(pygame.time.get_ticks()), audio_api)

        # Translate low level events to domain events
        while ev := pygame.event.poll():
            msg = None
            if ev.type == pygame.MOUSEBUTTONDOWN:
                if ev.button == 1:
                    msg = LeftMouseDownAt(ev.pos)
            if ev.type == pygame.MOUSEBUTTONUP:
                if ev.button == 1:
                    msg = LeftMouseUpAt(ev.pos)
            if ev.type == pygame.MOUSEMOTION:
                msg = MouseMovedTo(ev.pos)
            if ev.type == pygame.KEYDOWN:
                if ev.key == pygame.K_q:
                    return
            elif ev.type == pygame.QUIT:
                return

            if msg:
                model = update(model, msg, audio_api)

        # Display current model, if any change found
        if old_model_repr != print_model(model):
            view(model, drawing_api)

        pygame.display.update()
        clock.tick(FPS)


class DrawingAPI:
    def __init__(self, screen, resource_path):
        self.screen = screen
        self.image_dict = {}
        self.resource_path = resource_path
        self.font = pygame.font.Font(f'{self.resource_path}/font.ttf', 30)

    def draw_rectangle(self, center, size, color):
        r = pygame.Rect(0, 0, *size)
        r.center = center
        pygame.draw.rect(self.screen, color, r)

    def draw_disc(self, center, size, color):
        pygame.draw.circle(self.screen, color, center, size, size)

    def draw_text(self, center, text, color):
        text_surface = self.font.render(text, True, color)
        text_rect = text_surface.get_rect()
        text_rect.center = center
        self.screen.blit(text_surface, text_rect)

    def draw_image(self, center, name, dimension):
        if (name, dimension) not in self.image_dict:
            self.image_dict[(name, dimension)] = self.load_and_scale(name, dimension)
        image = self.image_dict[(name, dimension)]
        pos = center[0] - WIDTH // 2, center[1] - HEIGHT // 2
        self.screen.blit(image, pos)

    def load_and_scale(self, name, dimension):
        p = f'{self.resource_path}/{name}.png'
        image = pygame.image.load(p)
        return pygame.transform.scale(image, dimension)


class AudioAPI:
    def __init__(self, resource_path):
        self.resource_path = resource_path
        self.sound = {}

    def play_music(self, name):
        p = f'{self.resource_path}/{name}.ogg'
        pygame.mixer.music.load(p)
        pygame.mixer.music.play(-1)

    def stop_music(self):
        pygame.mixer.music.stop()

    def play_sound(self, name):
        if name not in self.sound:
            p = f'{self.resource_path}/{name}.wav'
            self.sound[name] = pygame.mixer.Sound(p)
        self.sound[name].play()


if __name__ == '__main__':
    main()

# TODO: all states as named tuples instead of classes and initial_GameState function
# TODO: refactor to many states instead of 'mixed in' states
# TODO: little man easter egg
# TODO: remove old_model_repr logic (animations on every screen so was premature optimization after all, but fun exp!)
# TODO: ALSA blocks mixer init. Can it be solved without resorting to 'catch exception / null audio api'?
# TODO: idea to reduce missing updating projectors with new state:
#         investigate using nametuple string representation in projectors


