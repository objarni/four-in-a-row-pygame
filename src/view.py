from src.constants import (CENTER, WIDTH, HEIGHT, Color, CENTER_X, CENTER_Y, DISC_DIAMETER, \
                           BOARD_LEFT, BOARD_RIGHT, DISC_RADIUS, BOARD_HEIGHT, DROP_DELAY_MS, COLUMNS, ROWS, EMPTY, RED)
from src.states import StartScreenState, GameState, GameOverState
from src.update import convert_to_column, positions_in_print_order, int_to_color_name


def view(model, drawing_api):
    clear_screen(drawing_api)
    if isinstance(model, StartScreenState):
        view_startscreenstate(model, drawing_api)
    if isinstance(model, GameState):
        view_gamestate(drawing_api, model)
    if isinstance(model, GameOverState):
        view_gameoverstate(drawing_api, model)


def view_startscreenstate(model, api):
    api.draw_image(CENTER, 'background', (WIDTH, HEIGHT))
    edged_text(api, "FOUR-IN-A-ROW", CENTER, Color.GREEN)
    edged_text(api, "Click left mouse button to play!",
               (CENTER_X, CENTER_Y + 30),
               Color.YELLOW)
    for i in range(200):
        api.draw_disc((int(model.time / 2 + i ** 2 * 37) % WIDTH, int(model.time + i * 237) % HEIGHT), 1, Color.BLUE)
    for i in range(4):
        bigger_disc = int(DISC_DIAMETER * 0.75)
        api.draw_disc((40 + i * (DISC_DIAMETER + 5), 100), bigger_disc, Color.YELLOW)
        api.draw_disc((WIDTH - 40 - i * (DISC_DIAMETER + 5), HEIGHT - 100), bigger_disc, Color.RED)


def frac(begin, end, current):
    return (float(current) - float(begin)) / (float(end) - float(begin))


assert frac(10, 20, 15) == 0.5
assert frac(10, 20, 10) == 0
assert frac(10, 20, 17.5) == 0.75


def view_gamestate(api, model):
    board = model.board
    draw_board(api, board)
    api.draw_text((WIDTH // 2, 20), f"{int_to_color_name(model.whos_turn_is_it).title()} to place disc", Color.WHITE)
    api.draw_rectangle((BOARD_LEFT, CENTER_Y), (2, 100), Color.GREEN)
    api.draw_rectangle((BOARD_RIGHT, CENTER_Y), (2, 100), Color.GREEN)

    # Player making move?
    if model.mouse_down_time:
        column = convert_to_column(model.mouse_down[0])
        pos = (BOARD_LEFT + DISC_DIAMETER * column + DISC_RADIUS, CENTER_Y - BOARD_HEIGHT // 2 - DISC_RADIUS)
        api.draw_disc(pos, DISC_RADIUS, rgb_from_color(model.whos_turn_is_it))
        # draw holding indicator
        fraction = frac(begin=model.mouse_down_time, end=model.mouse_down_time + DROP_DELAY_MS,
                        current=model.time)  # from, to, current
        api.draw_disc(pos, int(DISC_RADIUS * fraction), Color.GREEN)
    else:
        column = convert_to_column(model.mouse_pos[0])
        if column is not None:
            pos = (BOARD_LEFT + DISC_DIAMETER * column + DISC_RADIUS, CENTER_Y - BOARD_HEIGHT // 2 - DISC_RADIUS)
            api.draw_disc(pos, DISC_RADIUS, rgb_from_color(model.whos_turn_is_it))
        else:
            api.draw_disc(model.mouse_pos, DISC_RADIUS, rgb_from_color(model.whos_turn_is_it))


def view_gameoverstate(api, model):
    draw_board(api, model.board, scale=1.0)

    edged_text(api,
               "GAME OVER",
               (CENTER_X, CENTER_Y - 30),
               Color.WHITE)

    edged_text(api,
               f"{int_to_color_name(model.winner)} won!".upper(),
               ((CENTER_X), (CENTER_Y + 30)),
               rgb_from_color(model.winner))


def edged_text(api, txt, pos, color):
    (x, y) = pos
    api.draw_text(
        (x - 2, y - 2),
        txt,
        Color.WHITE
    )
    api.draw_text(
        (x + 2, y + 2),
        txt,
        Color.BLACK
    )
    api.draw_text(
        (x, y),
        txt,
        color
    )


def clear_screen(api):
    api.draw_rectangle(CENTER, (WIDTH, HEIGHT), Color.BLACK)


def draw_board(api, board, scale=1.0):
    api.draw_rectangle(CENTER, (DISC_DIAMETER * COLUMNS + 20, DISC_DIAMETER * ROWS + 20), Color.BLUE)
    for (x, y) in positions_in_print_order():
        value = board[(x, y)]
        color = Color.BLACK
        if value != EMPTY:
            color = rgb_from_color(value)
        x0 = int((WIDTH / 2 - COLUMNS * DISC_DIAMETER / 2 + x * DISC_DIAMETER + DISC_RADIUS))
        y0 = int((HEIGHT / 2 - ROWS * DISC_DIAMETER / 2 + y * DISC_DIAMETER + DISC_RADIUS))
        api.draw_disc((x0, y0), int(scale * DISC_RADIUS), color)


def rgb_from_color(color):
    return Color.RED if color == RED else Color.YELLOW
