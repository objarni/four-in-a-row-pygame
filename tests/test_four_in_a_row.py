
import pygame
import pytest
from approvaltests import verify

import src.constants
import src.four_in_a_row
import src.states
import src.update
from src.messages import LeftMouseDownAt, ColumnWasClicked, MouseMovedTo
from src.states import GameState, GameOverState, StartScreenState
from tests.fakes import simulate_main_event_loop
from tests.printers import print_rgb, print_scenario, SimulationLog


@pytest.fixture
def log():
    logger = SimulationLog()
    src.update.log = logger
    return logger


@pytest.fixture
def game_underway():
    return GameState()


def setup_function():
    pygame.init()


def test_print_rgb():
    assert print_rgb(0, 0, 0) == '.'
    assert print_rgb(200, 0, 0) == 'R'
    assert print_rgb(0, 200, 0) == 'G'
    assert print_rgb(0, 0, 200) == 'B'
    assert print_rgb(200, 200, 0) == 'Y'
    assert print_rgb(200, 200, 200) == 'W'


def test_first_placed_brick_is_red(log):
    game = GameState()
    player_moves = [ColumnWasClicked(0)]
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_placing_4_bricks_in_first_column(log):
    game = GameState()
    player_moves = [ColumnWasClicked(0)] * 4
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_placing_too_many_bricks_in_first_column(log):
    game = GameState()
    player_moves = [ColumnWasClicked(0)] * 8
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_letting_red_win(log):
    game = GameState()
    player_moves = [ColumnWasClicked(0), ColumnWasClicked(1)] * 3 + [ColumnWasClicked(0)]
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_letting_red_win_horisontally(log):
    game = GameState()
    player_moves = [ColumnWasClicked(c) for c in [0, 5, 1, 5, 2, 5, 3]]
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_letting_yellow_win_horisontally(log):
    game = GameState()
    player_moves = [ColumnWasClicked(c) for c in [0, 1, 0, 2, 0, 3, 1, 4]]
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_slash_red_win(log):
    game = GameState()
    player_moves = [ColumnWasClicked(c) for c in [0, 1, 1, 2, 2, 3, 2, 3, 3, 5, 3]]
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_backslash_yellow_win(log):
    game = GameState()
    player_moves = [ColumnWasClicked(c) for c in [0, 6, 5, 5, 4, 4, 3, 4, 5, 3, 0, 3, 0, 3]]
    result = simulate_main_event_loop(game, player_moves, log)
    verify(print_scenario(player_moves, result, log))


def test_startscreen(log):
    game = StartScreenState()
    ticks = [src.messages.Tick(ms) for ms in range(3)]
    result = simulate_main_event_loop(game, ticks, log)
    verify(print_scenario(ticks, result, log))


def test_clicking_in_game_over_state(log):
    game = GameOverState(winner=src.constants.RED, board=src.states.empty_board())
    mouse_events = [LeftMouseDownAt((1, 1))]
    result = simulate_main_event_loop(game, mouse_events, log)
    verify(print_scenario(mouse_events, result, log))


def test_startscreen_to_game_transition(log):
    game = StartScreenState()
    mouse_events = [LeftMouseDownAt((1, 1))]
    result = simulate_main_event_loop(game, mouse_events, log)
    verify(print_scenario(mouse_events, result, log))


def test_mouse_movement_over_game_screen(log):
    game = GameState()
    mouse_events = [MouseMovedTo((x, x)) for x in range(0, 500, 10)]
    result = simulate_main_event_loop(game, mouse_events, log)
    verify(print_scenario(mouse_events, result, log))
