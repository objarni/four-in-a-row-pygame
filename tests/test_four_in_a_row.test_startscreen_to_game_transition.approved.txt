SCENARIO:
startscreen_to_game_transition
events:
LeftMouseDownAt(pos=(1, 1))


FINAL STATE:
GameState
It is reds turn.
model.time=None
The mouse is at (550, 400).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O


FINAL SCREEN:
...............................................................................
...............................W..W...........W................................
..............................WWWWW.WWWW.WWW.W.WWW.............................
.......................................RR......................................
......................................RRRR.....................................
.....................................RRRRRR....................................
.....................................RRRRRR....................................
.....................................RRRRRR....................................
.....................................RRRRRR....................................
.................BBBBBBBBBBBBBBBBBBBBBBRRBBBBBBBBBBBBBBBBBBBBBB................
.................BB....BB....BBB....BB....BB....BBB....BB....BB................
.................B.....B......B.....B......B.....B......B.....B................
...............................................................................
...............................................................................
.................B.....B......B.....B......B.....B......B.....B................
.................BB....BB....BBB...BBB....BBB...BBB....BB....BB................
.................BB...BBBB..BBBB...BBBB..BBBB...BBBB..BBBB...BB................
.................B.....B......B.....B......B.....B......B.....B................
.................B............................................B................
...............................................................................
.................B............B..................B............B................
.................B.....BB.....B.....BB....BB.....B.....BB.....B................
.................BBB.BBBBB..BBBBB.BBBBB..BBBBB.BBBBB..BBBBB.BBB................
.................B.....BB....BB.....BB....BB.....BB....BB.....B................
.................B............B.....B......B.....B............B................
...............................................................................
...............................................................................
.................B.....B......B.....B......B.....B......B.....B................
.................BB...BBB....BBB...BBB....BBB...BBB....BBB...BB................
.................BB...BBB....BBB...BBB....BBB...BBB....BBB...BB................
.................B.....B......B.....B......B.....B......B.....B................
...............................................................................
...............................................................................
.................B............B.....B......B.....B............B................
.................B.....BB....BB.....BB....BB.....BB....BB.....B................
.................BBB.BBBBB..BBBBB.BBBBB..BBBBB.BBBBB..BBBBB.BBB................
.................B.....BB.....B.....BB....BB.....B.....BB.....B................
.................B............B..................B............B................
...............................................................................
.................B............................................B................
.................B.....B......B.....B......B.....B......B.....B................
.................BB...BBBB..BBBB...BBBB..BBBB...BBBB..BBBB...BB................
.................BB....BB....BBB...BBB....BBB...BBB....BB....BB................
.................B.....B......B.....B......B.....B......B.....B................
...............................................................................
...............................................................................
.................B.....B......B.....B......B.....B......B.....B................
.................BB....BB....BBB....BB....BB....BBB....BB....BB................
.................BBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBBB................
...............................................................................
...............................................................................
...............................................................................
...............................................................................
...............................................................................
...............................................................................
...............................................................................
...............................................................................


SIMULATION LOG:
[SIMULATION STARTING]
===Model state===
StartScreenState
model.time=0
model.music_playing=False



[SIMULATING MSG=LeftMouseDownAt(pos=(1, 1))]

Stopping music playback.
===Model state===
GameState
It is reds turn.
model.time=None
The mouse is at (550, 400).
model.mouse_down_time=None
0 1 2 3 4 5 6
-------------
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O
O O O O O O O



[SIMULATION ENDED]
